pkg/wireguard/agent/agent.go

```diff
@@ -174,73 +193,102 @@ func (a *Agent) RestoreFinished() error {
 	return nil
 }
 
-func (a *Agent) UpdatePeer(nodeName, pubKeyHex string,
-	nodeIPv4 net.IP, podCIDRv4 *net.IPNet,
-	nodeIPv6 net.IP, podCIDRv6 *net.IPNet) error {
+func (a *Agent) UpdatePeer(nodeName, pubKeyHex string, nodeIPv4, nodeIPv6 net.IP) error {
+	// To avoid running into a deadlock, we need to lock the IPCache before
+	// calling a.Lock(), because IPCache might try to call into
+	// OnIPIdentityCacheChange concurrently
+	a.ipCache.RLock()
+	defer a.ipCache.RUnlock()
 
 	a.Lock()
 	defer a.Unlock()
 
-	// Handle pubKey change
-	if prevPubKeyHex, found := a.pubKeyByNodeName[nodeName]; found && prevPubKeyHex != pubKeyHex {
-		log.WithField(logfields.NodeName, nodeName).Info("Pubkey has changed")
-		// pubKeys differ, so delete old peer
-		if err := a.deletePeerByPubKey(prevPubKeyHex); err != nil {
-			return err
+	var allowedIPs []net.IPNet = nil
+	if prev := a.peerByNodeName[nodeName]; prev != nil {
+		// Handle pubKey change
+		if prev.pubKey.String() != pubKeyHex {
+			log.WithField(logfields.NodeName, nodeName).Info("Pubkey has changed")
+			// pubKeys differ, so delete old peer
+			if err := a.deletePeerByPubKey(prev.pubKey); err != nil {
+				return err
+			}
 		}
-		delete(a.pubKeyByNodeName, nodeName)
-	}
 
-	log.WithFields(logrus.Fields{
-		logfields.NodeName:  nodeName,
-		logfields.PubKey:    pubKeyHex,
-		logfields.NodeIPv4:  nodeIPv4,
-		logfields.PodCIDRv4: podCIDRv4,
-		logfields.NodeIPv6:  nodeIPv6,
-		logfields.PodCIDRv6: podCIDRv6,
-	}).Info("Adding peer")
+		// Reuse allowedIPs from existing peer config
+		allowedIPs = prev.allowedIPs
 
-	pubKey, err := wgtypes.ParseKey(pubKeyHex)
-	if err != nil {
-		return err
+		// Handle Node IP change
+		if !prev.nodeIPv4.Equal(nodeIPv4) {
+			delete(a.nodeNameByNodeIP, prev.nodeIPv4.String())
+			allowedIPs = nil // reset allowedIPs and re-initialize below
+		}
+		if !prev.nodeIPv6.Equal(nodeIPv6) {
+			delete(a.nodeNameByNodeIP, prev.nodeIPv6.String())
+			allowedIPs = nil // reset allowedIPs and re-initialize below
+		}
 	}
 
-	allowedIPs := []net.IPNet{}
-
-	if option.Config.EnableIPv4 {
-		if podCIDRv4 != nil {
-			allowedIPs = append(allowedIPs, *podCIDRv4)
+	if allowedIPs == nil {
+		// (Re-)Initialize the allowedIPs list by querying the IPCache. The
+		// allowedIPs will be updated by OnIPIdentityCacheChange after this
+		// function returns.
+		var lookupIPv4, lookupIPv6 net.IP
+		if option.Config.EnableIPv4 && nodeIPv4 != nil {
+			lookupIPv4 = nodeIPv4
+			allowedIPs = append(allowedIPs, *iputil.IPToPrefix(nodeIPv4))
 		}
-	}
-	if option.Config.EnableIPv6 {
-		if podCIDRv6 != nil {
-			allowedIPs = append(allowedIPs, *podCIDRv6)
+		if option.Config.EnableIPv6 && nodeIPv6 != nil {
+			lookupIPv6 = nodeIPv6
+			allowedIPs = append(allowedIPs, *iputil.IPToPrefix(nodeIPv6))
 		}
+		allowedIPs = append(allowedIPs, a.ipCache.LookupByHostRLocked(lookupIPv4, lookupIPv6)...)
+	}
+
+	pubKey, err := wgtypes.ParseKey(pubKeyHex)
```

Comment:
I think we check the pubKey too late. I've just seen:
```
level=warning msg="Failed to update wireguard configuration for peer" error="wgtypes: incorrect key size: 0" nodeName=cilium3 subsys=linux-datapath
```

My suggestion is to return early if `pubKeyHex == ""`.

Reply:
I can confirm this. I decided to fix this at the call site however. If a node is not ready to be added to the Wireguard pool, `UpdatePeer` should not be called imo.

https://github.com/cilium/cilium/blob/fb48284aadd2ccd756ea50cf8bcd1fac6705e3bf/pkg/datapath/linux/node.go#L926-L927